src/oemof/solph/flows/_non_convex_invest_flow.py

```diff
@@ -0,0 +1,849 @@
+        A dictionary containing the following two keys:
+         * `'ub'`: numeric (iterable, scalar or None), the normed *upper
+           bound* on the positive difference (`flow[t-1] < flow[t]`) of
+           two consecutive flow values.
+         * `'costs``: numeric (scalar or None), the gradient cost per
+           unit.
+    negative_gradient : :obj:`dict`, default: `{'ub': None, 'costs': 0}`
+        A dictionary containing the following two keys:
+          * `'ub'`: numeric (iterable, scalar or None), the normed *upper
+            bound* on the negative difference (`flow[t-1] > flow[t]`) of
+            two consecutive flow values.
+          * `'costs``: numeric (scalar or None), the gradient cost per
+            unit.
+    """
+
+    def __init__(
+        self,
+        startup_costs=None,
+        shutdown_costs=None,
+        activity_costs=None,
+        inactivity_costs=None,
+        minimum_uptime=None,
+        minimum_downtime=None,
+        maximum_startups=None,
+        maximum_shutdowns=None,
+        initial_status=0,
+        positive_gradient=None,
+        negative_gradient=None,
+        **kwargs,
+    ):
+        default_gradient = {"ub": None, "costs": 0}
+        if positive_gradient is None:
+            positive_gradient = default_gradient
+        if negative_gradient is None:
+            negative_gradient = default_gradient
+        nonconvex = NonConvex(
+            startup_costs=startup_costs,
+            shutdown_costs=shutdown_costs,
+            activity_costs=activity_costs,
+            inactivity_costs=inactivity_costs,
+            minimum_uptime=minimum_uptime,
+            minimum_downtime=minimum_downtime,
+            maximum_startups=maximum_startups,
+            maximum_shutdowns=maximum_shutdowns,
+            initial_status=initial_status,
+            positive_gradient=positive_gradient,
+            negative_gradient=negative_gradient,
+        )
+
+        other_warning_messages = []
+        with warnings.catch_warnings(record=True) as w:
+            # Only in the NonConvexInvestClass, the attribute
+            # `allow_nonconvex_investment` can be set to True.
+            super().__init__(
+                nonconvex=nonconvex, allow_nonconvex_investment=True, **kwargs
+            )
+
+            for wm in w:
+                if isinstance(wm.message, FlowOptionWarning):
+                    continue
+                else:
+                    other_warning_messages.append(wm.message)
+
+        for wm in other_warning_messages:
+            warnings.warn(wm)
+
+        # If an <class 'solph.Investment'> instance is used then its
+        # 'nonconvex' attribute should be set to True and the checks
+        # should be run again
+        if self.investment is not None:
+            self.investment.nonconvex = True
+            self.investment._check_invest_attributes_maximum()
+            self.investment._check_invest_attributes()
```

Comment:
I do not understand why these checks are repeated and the `self.investment.nonconvex` is set True. This attribute was intended to add a investment with fixed and capacity dependent investment costs (see [https://oemof-solph.readthedocs.io/en/latest/reference/oemof.solph.html#oemof.solph.flows._investment_flow.InvestmentFlowBlock](https://oemof-solph.readthedocs.io/en/latest/reference/oemof.solph.html#oemof.solph.flows._investment_flow.InvestmentFlowBlock))

Thus, I would rather recommend to remove this lines:

```suggestion
```

You could put a warning if either `investment.nonconvex` is True or `investment.offset` is given.

Actually, an investment attriubte is required for the NonConvexInvestFlow.

But maybe it makes sense to resolve the `Investment` class in the flow classes -- wasn't this decided on some dev meeting? Then, all this kind of issues become cleaner.

Reply:
In the end, this kind of combination of a nonconvex Investment (having a minimum investment threshold with an fixed offset value) and your NonConvexInvestFlow is not implemented, and the user should know about that.

Reply:
This was to be able to raise an AttributeError if the `solph.Investment` instance passed as init argument of `solph.flows.NonConvexInvestFlow` does not have `maximum` provided, the test test_nonconvex_investment_without_maximum_raises_warning was written to make sure an error is triggered. Said otherwise there would be nothing which prevents the user to provide a convex `solph.Investment`. This would not make sense, so we force the `nonconvex` of the `solph.Investment` instance to `True` and rerun the checks that would have only run if the `solph.Investment` instance would have had `nonconvex` set to `False`. The test [test_nonconvex_investment_without_maximum_raises_warning](https://github.com/sasa821/oemof-solph/blob/cc747bd6a770b23020b795374762e96e89417db1/tests/test_warnings.py#L157) we added would fail if you comment those lines.

Reply:
ah okay, I see the point, the `maximum` is required for the NonConvexInvestFlow and you want to use the check. However, in my opinion, there should be some adaptions, the usage of the Investment attributes should be consistent with the nonconvex option of the InvestmentFlow, and therefore (maybe until a better solution is found), the checks should be adapted specifically to the NonConvexInvestFlow:
 * the `Investment` class/attribute needs to be there at all (otherwise, the NonConvexInvestFlow does not work)
 * `ep_costs` needs to be given
 * `maximum` needs to be given
 * `offset` must not be given (as it does not have an effect)
 * `existing` must not be given (as it does not have an effect)
 * `nonconvex` must be `False` (otherwise, the meaning of the attributes of the Investment class is not consistent)
 * `minimum` is optional (and sets a "hard" minimum which means that the investment cannot become zero)

Reply:
I agree with you, this would actually be easy to implement once the `get rid of "**kwargs"` todo has been implemented (we agreed above to tackle that in a subsequent PR), I suggest we add your comment as a TODO inside the code